# Introdução a Python

(este Notebook está em permanente construção)

## Python

Quando comecei na área de engenharia o computador não era tão difundido e a calculadora científica programável era a ferramenta indispensável do engenheiro para a realização dos cálculos. Hoje é impensável engenharia sem a utilização do computador como ferramenta. Se antes tínhamos que aprender a usar a calculadora científica, hoje temos que aprender a usar aplicativos de smartphone, planilhas, e mais recentemente, uma ferramenta poderosa para usar nos projetos de engenharia, que é o ambiente Python.

Python não é apenas uma linguagem de programação, ele é um ecosistema constituído de vários recursos, como bibliotecas de programas para solução de problemas diversos, além de recursos como o Google Colab e o Jupyter Notebook, que estou utilizanddo para escrever este documento. 

Não há necessidade de ser um programador para utilizar Python. Na verdade não é necessário nem escrever um único programa para utilizar Python. Pode-se começar pelo básico na linha de comando e já obter algum resultado. Para quem conhece calculadoras científicas programáveis, é como se estivéssemos usando o teclado da calculadora sem usar o modo de programação.

Python é interativo e fácil de usar, e durante a utilização ele geralmente tenta mostrar algum tipo de resultado ao invés de uma mensagem de erro, como aconteceria em outras "linguagens". Isso facilita bastante o aprendizado, pode-se "brincar" à vontade, ir digitando quase como que "psicografando" e ir obtendo resultados parciais até chegar no resultado final.

Em particular, quando combinado com recursos como Jupyter Notebook (assim como este documento que estou escrevendo), além de bibliotecas prontas, Python torna-se uma das ferramentas mais poderosas que temos atualmente para utilizar na engenharia.

A intenção deste documento não é ensinar programação, nem ser abrangente ou completo. O objetivo é mostrar como solucionar problemas simples de engenharia utilizando conceitos básicos de Python. Sugiro utilizar o [Google Colab](https://colab.research.google.com/) para executar os exemplos a seguir, porém existem outras formas de executá-los.

## Calculadora

Vamos começar com um exemplo extraído do manual da calculadora HP-34C, adaptado:

![](img/ganimedes.jpg)

Ganimedes, uma das 12 luas de Júpiter, tem um diâmetro de 3200 milhas. Você pode usar Python para calcular a área da superfície de Ganimedes, passo a passo:

In [1]:
D1 = 3200.0    # diâmetro de Ganimedes
D1             # mostra o diâmetro, armazenado na variável D

3200.0

O que está após o jogo da velha (`#`) são comentários, e não é executado. Serve apenas para documentar o que está sendo feito.

Na primeira linha, `D1` é o nome da variável aonde armazenamos o número `3200.0`. Note que ele tem o ponto decimal seguido de zero (`.0`) que é para indicar que é um número real de número flutuante. Sem o `.0` seria um número inteiro.
Na segunda linha, o `D1` sozinho indica que é para mostrar o valor que está na variável, o que é feito na linha de baixo, o `3200.0`. Notar mais uma vez o `.0` no final do número indicando que é um número real de ponto flutuante.

A área de uma superfície esférica pode ser calculada usando a fórmula &nbsp;&nbsp; A = &pi; &times; D &sup2;

O número Pi (&pi;) pode ser obitido de diversas maneiras. Pode-se simplesmente digitar "3.1415" que talvez seja o mais fácil, ou então utilizar uma constante definida em uma biblioteca:

In [2]:
import math         # biblioteca
math.pi             # mostra o número Pi

3.141592653589793

Na primeira linha o `import` chama uma biblioteca, que nada mais é do que um conjuntos de programas, constantes, e outras coisas que alguém já fez e deixou pronto para usar na biblioteca `math`. Acho que as bibliotecas são a parte mais importante do Python, pois tem muita gente fazendo bibliotecas e deixando pronto para as outras pessoas usarem, tem biblioteca de tudo que se possa imaginar.

Neste nosso caso alguém já defininiu a constante `pi` e deixou pronto na biblioteca `math`. Se fosse uma calculadora científica, existiria uma tecla "&pi;" que bastaria apertar e o número Pi apareceria no visor, ou seja, alguém já deixou pronto o número Pi na calculadora e para mostrar basta apertar a tecla apropriada. Algumas calculadoras científicas programáveis permitem redefinir teclas, de tal forma que se não tiver a tecla Pi pré-programada, basta fazer um programa na calculadora que mostra o número Pi, e associar o programa a uma tecla. Ao apertar a tecla o número Pi aparece. O conceito aqui é programar a constante e deixar disponível para quando for usar. Mais para a frente falaremos mais sobre as bibliotecas Python. No momento não vamos programar nada, vamos só usar o que já está pronto.

Na segunda linha, `math.pi` é a forma como chamamos o número `pi` que está na biblioteca `math`.

Continuando:

In [3]:
A1 = D1**2 * math.pi    # na variável A1 colocamos o resultado do cálculo
A1                      # o A1 sozinho, mostra o resultado

32169908.772759482

O sinal `**` é o de exponenciação, assim `D1**2` eleva o conteúdo da variável `D1` ao quadrado.

O `*` é a multiplicação, e `math.pi` é o número Pi que alguém já colocou na biblioteca `math` e basta usar.

Na variável `A1` armazenamos o resultado `32169908.772759482` que é a área da superfície de Ganimedes em milhas quadradas. É interessante ir colocando os resultados intermediários em variáveis para usar em cálculos subsequentes sem ter que redigitar. No caso colocamos o resultado do cálculo da superfície na variável `A1`. Se fosse uma calculadora científica, colocaríamos em uma memória, por exemplo, `STO 01`.

Uma curiosidade aqui é que o resultado que está no manual da calculadora HP-34C é `32,169,908.78` .


Io, outra lua de Júpiter, tem 3643.2 km de diâmetro. A área da superfície dela é maior ou menor do que a de Ganimedes?

In [13]:
D2 = 3643.2             # diãmetro de Io, em kilômetros
A2 = D2**2 * math.pi    # mesma biblioteca "math" usada no cálculo acima
A2                      # mostra a área, em kilômetros quadrados

41698064.73537012

Agora basta comparar as duas áreas, certo?

Só tem um probleminha, uma está em milhas quadradas, e a outra em kilômetros quadrados, então precisamos converter. Para converter de milhas para kilômetros multiplicamos por 1.60934. Talvez seja mais fácil refazer os cálculos, pois usando Python e Jupyter Notebook é fácil fazer um Ctrl-C/Ctrl-V:

In [19]:
D1 = 3200 * 1.60934     # conversão de milhas para kilômetros
A1 = D1**2 * math.pi    # área em kilômetros quadrados
A1                      # mostra o resultado

83319267.05295825

Agora sim podemos comparar. Porém quando os números são grandes assim, a comparação visual fica um pouco difícil. Como temos as áreas armazenadas em variáveis (A1 e A2) talvez seja mais fácil perguntar para o Python:

In [21]:
A1 > A2

True

Ou seja, a área de Ganimedes (A1) é maior do que a de Io (A2).

Agora vamos perguntar se a área de Io é maior que a de Ganimedes:

In [23]:
A2 > A1

False

Não, não é.

E podemos brincar à vontade com as perguntas:

In [32]:
A1 == A2        # as áreas são iguais?

False

In [28]:
A1 != A2        # são diferentes?

True

In [31]:
A1 < A2         # é menor?

False

In [34]:
A1 > (2 * A2)   # Ganimedes é maior que o dobro de Io ?

False

Note que até o momento não usamos uma linha sequer de programação, e assim mesmo já obtivemos vários resultados. Essa interatividade e facilidade de uso são coisas que atraem em Python.